text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,1275 @@
+as well and this is implemented in a nightly compiler right now.
+For example, we can make and invoke macro that ascribes an expression with a
+type `Vec<$t>` with the following valid snippet:
+
+```rust
+#![feature(type_ascription)]
+
+macro_rules! ascribe {
+    ($e: expr, $t: ty) => {
+        $e : Vec<$t>
+    }
+}
+
+fn main() {
+    let _ = ascribe!(vec![1, 2, 3], u8);
+}
+```
+
+### Precedence of the operator
+
+[prec_unresolved]: https://github.com/rust-lang/rfcs/blob/master/text/0803-type-ascription.md#unresolved-questions
+
+[prec_ref]: https://doc.rust-lang.org/beta/reference/expressions.html
+
+[RFC 803] proposed and implemented that `:` as an operator in expression
+contexts should have the same precedence as the `as` operator
+*([see the reference][prec_ref])*. However, the RFC also
+[left this question unresolved][prec_unresolved] and asked:
+
+> Is the suggested precedence correct?
+
+We argue in this RFC that the current implementation is sub-optimal and thus
+propose that the precedence should be slightly changed.
+
+To see why, consider the example above where we wrote:
+
+```rust
+let _ = ("1234".parse().unwrap() : usize).into() : Box<_>;
+```
+
+Notice in particular here that we've had to enclose the inner ascription
+in parenthesis. Consider that you are writing this snippet and reach the
+`.into()`. Once you do that, you'll need to select everything on the line
+until before the `=` token. This can slow down your writing flow.
+Furthermore, as we chain more and more methods, the build-up of parenthesis
+can increase and thus make writing and reading further impaired. An example:
+
+```rust
+let x = (((0..10)
+    .map(some_computation)
+    .collect() : Result<Vec<_>, _>)
+    .unwrap()
+    .map(other_computation) : Vec<usize>)
+    .into() : Rc<[_]>;
+```
+
+We suggest instead that you should be able to write:
+
+```rust
+let x = (0..10)
+    .map(some_computation)
+    .collect() : Result<Vec<_>, _>
+    .unwrap()
+    .map(other_computation) : Vec<usize>
+    .into() : Rc<[_]>;
+```
+
+To that end, `foo : bar.quux()` and `foo : bar.quux` should unambiguously be
+interpreted as `(foo : bar).quux()` and `(foo : bar).quux`. 
+
+However, this does not mean that the operator `:` should bind more tightly than
+operators such as the unary operators `-`, `*`, `!`, `&`, and `&mut`.
+In particular, for the latter two operators, we expect that if someone writes
+`&x : Type`, it would be interpreted as `(&x) : Type` as opposed to `&(x : Type)`.
+
+Instead, we propose that whenever type ascription is followed by a
+field projection or a method call, the projections or the call should apply
+to the entire ascribed expression.
+
+Note in particular that when you write `&a:b.c`, because `&` binds more tightly
+than `:` but `.` binds more tightly than `&`, the expression associates as
+`&((a : b).c)`. However, when you write `&x.y:z`, it instead associates as
```

Comment:
Your further specification says:

> when the parser encounters type ascription of an expression immediately followed by a field projection or a method call, then the parser shall interpret the projection and the call as being performed on the ascribed expression

Since the _ascribed expression_ in `&a:b.c` is `&a`, it seems (looking at your own specification) that this should rather be parsed as `((&a) : b).c`. Parsing it differently would also be very confusing, because

* `&a:b` is parsed as `(&a) : b`
* `&a:b.c` _would be_ parsed as `&((a : b).c)`

and hence adding the postfix `.c` _would change_ how the postfixed tokens are parsed. This detail would diminish the reading/writing flow. Such a rule therefore contradicts your other arguments about the flow.

### Alternative

I propose to instead parse this in the way I interpreted your specification: `((&a) : b).c`. This admittedly has a similar problem:

* `&a.c` is parsed as `&(a.c)`
* `&a:b.c` is parsed as `((&a) : b).c`

(I.e. inserting the `:b` in the middle makes the context of the `&`-operator more local.)

With this rule however, the change that introduces the inconsistency has to be applied in the middle. The flow therefore isn't diminished if one reads/writes from left to right. Also, when applying a change in the middle, it's probably more expected by the writer that the semantics change, compared to when just appending something.

Your exampes should still work.